### PR TITLE
SER-3442 Track search result clicks on community links 

### PR DESCRIPTION
**Pull request by mateuszsip**

### CODE REVIEW

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
+	};
+
+	var getUniqueSearchId = function() {
+		if (this.searchUID) {
+			return this.searchUID;
+		}
+
+		var queryParams = new URL(window.location).searchParams;
+		var searchUID = queryParams.get('searchUID') || uuid();
+		this.searchUID = searchUID;
```
Comment (mmeller-wikia) [inline extensions/wikia/Search/js/WikiaSearch.js]: Can we assign searchUID directly to this.searchUID and after that return it?

**Review by mmeller-wikia — APPROVED** [review]

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 2]
```diff
@@ -1,6 +1,54 @@
 (function($, window) {
 
 require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTracking, uuid, trackingOptIn) {
+	var resultClickTrackerFactory = function (type, idGenerator, filtersProvider) {
+		return function(clickedElement) {
+			var queryparams = new URL(window.location).searchParams;
+			var query = queryparams.get('search') || queryparams.get('query');
+
+			if (!query) {
+				return;
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js]: I'm not sure this is right, because on the search page, the users can enter an empty query, and clicking something would mean that the user got something valuable, right?

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 3]
```diff
@@ -91,37 +147,38 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
...
-			if (!query) {
-				return;
+		trackSearchResultClick: resultClickTrackerFactory(
+			'article',
+			function (clickedElement) {
+				return clickedElement.getAttribute('data-wiki-id') + '_' + clickedElement.getAttribute('data-page-id');
+			},
+			function () {
+				return {
+					searchType: getCurrentScope()
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js]: Why is this a lambda? Can current scope change dynamically? I thought it's a page refresh.

**Review by slayful — CHANGES_REQUESTED** [review]
Few questions and nitpicks, besides that LGTM.

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 2 above)
Comment (mateuszsip) [reply extensions/wikia/Search/js/WikiaSearch.js]: It was like that (probably because old search had similar condition on backend), but you're right - I will change that.
Thanks!

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 1 above)
Comment (mateuszsip) [reply extensions/wikia/Search/js/WikiaSearch.js]: Yes, we can - I was extracting old code without refactoring, but I will change that. Thanks!

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 3 above)
Comment (mateuszsip) [reply extensions/wikia/Search/js/WikiaSearch.js]: You are right, we didn't think about it when scope was introduced. I will change that, thanks!

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 4]
```diff
@@ -33,20 +35,14 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
...
 		if (this.searchUID) {
 			return this.searchUID;
 		}
 
 		var queryParams = new URL(window.location).searchParams;
-		var searchUID = queryParams.get('searchUID') || uuid();
-		this.searchUID = searchUID;
 
-		return searchUID;
+		return queryParams.get('searchUID') || uuid();
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js]: If searchUID is not available shouldn't you generate it once and assingn to `this.SearchUID`?

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 5]
```diff
@@ -154,7 +150,7 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
 			},
 			function () {
 				return {
-					searchType: getCurrentScope()
+					searchType: currentScope
 				};
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js]: I meant that this parameter doesn't need to be a function, `resultClickTrackerFactory` can be modified.

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 2 above)
Comment (slayful) [reply extensions/wikia/Search/js/WikiaSearch.js]: So, will ya? :)

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 4 above)
Comment (mateuszsip) [reply extensions/wikia/Search/js/WikiaSearch.js]: Good catch - it's not generated backend side, so for first search page id for impression and click tracking is always different 🤔
This is old code, so it could be broken for a long time

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 6]
```diff
@@ -1,6 +1,46 @@
...
+			trackingOptIn.pushToUserConsentQueue(function () {
+				window.searchTracking.trackSearchClicked(payload);
+			});
+		}.bind(this);
+	};
+
+	var getUniqueSearchId = function() {
+		if (!this.searchUID) {
+			var queryParams = new URL(window.location).searchParams;
+
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js:37]:
```suggestion
```

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 7]
```diff
@@ -36,10 +76,18 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
 				}
 			});
 
-			$('.result-link').on('click', function(event) {
+			$('.result h1 .result-link').on('click', function(event) {
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js]: The selector seems a bit fragile. How about we use a different class for the community link?

Let's also make sure how is this code related to `app/skins/oasis/js/Tracking.js`.

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 8]
```diff
@@ -91,53 +139,42 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
...
-		trackSearchResultClick: function(clickedElement) {
-			var queryparams = new URL(window.location).searchParams;
-			var query = queryparams.get('search') || queryparams.get('query');
-
-			if (!query) {
-				return;
+		trackSearchResultClick: resultClickTrackerFactory(
+			'article',
+			function (clickedElement) {
+				return clickedElement.getAttribute('data-wiki-id') + '_' + clickedElement.getAttribute('data-page-id');
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js:145]:
```suggestion
				return clickedElement.data('wiki-id') + '_' + clickedElement.data('page-id');
```

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 9]
```diff
@@ -91,53 +139,42 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
...
+			},
+			{}
+		),
+		trackSearchResultCommunityClick: resultClickTrackerFactory(
+			'community',
+			function (clickedElement) {
+				return clickedElement.getAttribute('data-wiki-id');
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js:152]:
```suggestion
				return clickedElement.data('wiki-id');
```

File: extensions/wikia/Search/js/WikiaSearch.js [hunk 10]
```diff
@@ -91,53 +139,42 @@ require(['search-tracking', 'uuid', 'wikia.trackingOptIn'], function(searchTrack
...
-
-			trackingOptIn.pushToUserConsentQueue(function () {
-				window.searchTracking.trackSearchClicked(payload);
-			});
-		},
+		),
+		trackRightRailResultClick: resultClickTrackerFactory(
+			'community',
+			function (clickedElement) {
+				return clickedElement.getAttribute('data-wiki-id');
```
Comment (slayful) [inline extensions/wikia/Search/js/WikiaSearch.js:161]:
```suggestion
				return clickedElement.data('wiki-id');
```

File: extensions/wikia/Search/js/WikiaSearch.js (hunk 8 above)
Comment (KockaAdmiralac) [reply extensions/wikia/Search/js/WikiaSearch.js:145]: Would that work? Parameter in a click event's callback is a plain [Node](https://developer.mozilla.org/en-US/docs/Web/API/Node) instance whereas [`.data`](https://api.jquery.com/data/) is a jQuery function that won't work on plain nodes (and isn't necessarily about getting data attribute values, it gets arbitrary data about that node that falls back to `data-*` attribute values if otherwise not found).

Comment (slayful) [reply extensions/wikia/Search/js/WikiaSearch.js:145]: You're right, it doesn't. :)

**Review by slayful — APPROVED** [review]
